feature-libs/product-configurator/rulebased/root/interceptor/cpq-access-storage.service.ts

```diff
@@ -0,0 +1,132 @@
+import { Injectable } from '@angular/core';
+import { AuthService } from '@spartacus/core';
+import { BehaviorSubject, Observable, Subscription, timer } from 'rxjs';
+import {
+  distinctUntilChanged,
+  expand,
+  filter,
+  switchMap,
+  take,
+} from 'rxjs/operators';
+import { CpqAccessData } from './cpq-access-data.models';
+import { CpqAccessLoaderService } from './cpq-access-loader.service';
+import { CpqConfiguratorAuthConfig } from './cpq-configurator-auth.config';
+
+@Injectable({ providedIn: 'root' })
+export class CpqAccessStorageService {
+  protected readonly EXPIRED_TOKEN = {
+    accessToken: 'INVALID DUMMY',
+    accessTokenExpirationTime: 0,
+  };
+
+  constructor(
+    protected cpqAccessLoaderService: CpqAccessLoaderService,
+    protected authService: AuthService,
+    protected config: CpqConfiguratorAuthConfig
+  ) {}
+
+  protected cpqAccessObservable: Observable<CpqAccessData>;
+  protected currentCpqAccessSubscription: Subscription;
+  protected cpqAccessDataCache: BehaviorSubject<CpqAccessData>;
+
+  protected currentCpqSessionId: string | null = null;
+
+  getCachedCpqAccessData(): Observable<CpqAccessData> {
+    if (!this.cpqAccessObservable) {
+      this.initCpqAccessObservable();
+    }
+    return this.cpqAccessObservable;
+  }
+
+  /**
+   * Renews the current access data. All subscribers of getCachedCpqAccessData()
+   * will receive the new data. Will only have an effect, if there are any subscribers
+   * and the user is logged in.
+   */
+  renewCachedCpqAccessData() {
+    // only force token refresh if initialized.
+    if (this.cpqAccessObservable) {
+      this.stopAutoFetchingCpqAccessData();
+      this.cpqAccessDataCache.next(this.EXPIRED_TOKEN); // invalidate cache
+      this.authService
+        .isUserLoggedIn()
+        .pipe(take(1)) // get current login state
+        .subscribe((loggedIn) => {
+          // only fetch new token if user is logged in.
+          if (loggedIn) {
+            this.startAutoFetchingCpqAccessData();
+          }
+        });
+    }
+  }
+
+  protected initCpqAccessObservable() {
+    this.cpqAccessDataCache = new BehaviorSubject(this.EXPIRED_TOKEN);
+    this.cpqAccessObservable = this.cpqAccessDataCache.pipe(
+      // Never expose expired tokens - either cache was invalidated with expired token,
+      // or the cached one expired before a new one was fetched.
+      filter((data) => !this.isTokenExpired(data))
+    );
+    this.authService
+      .isUserLoggedIn()
+      .pipe(distinctUntilChanged()) // only react if user login status changes
+      .subscribe((loggedIn) => {
```

Comment:
[nitpick] It's a good practice to dispose all long living subscriptions in `ngOnDestroy` **also for services** to avoid memory leaks in SSR. In the long-living NodeJS process of the SSR server the pending subscriptions survive even after the bootstrapped app is destroyed.

However your code is not executed in SSR (becasue your feature requires prior login), so this comment is only a nitpick.

Reply:
pick